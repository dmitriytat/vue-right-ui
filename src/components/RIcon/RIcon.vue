<template>
  <svg
    :width="dimensions.width"
    :height="dimensions.height"
    :viewBox="glyph.viewBox"
    class="r-icon"
    :aria-hidden="!standalone"
    :role="standalone ? 'img' : undefined"
  >
    <use :xlink:href="glyph.url" />
  </svg>
</template>

<script>
import sizeMixin from '@/mixins/sizeMixin';

const SIZES = {
  small: {
    width: 8,
    height: 8,
  },

  medium: {
    width: 12,
    height: 12,
  },

  large: {
    width: 16,
    height: 16,
  },
};

export default {
  name: 'RIcon',

  mixins: [sizeMixin],

  props: {
    /**
     * Svg icon
     */
    glyph: {
      type: Object,
      required: true,
    },

    /**
     * Icon is used without text.
     * Probably you need to use aria-label to describe it.
     */
    standalone: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    dimensions() {
      return SIZES[this.size];
    },
  },
};
</script>

<style>
  .r-icon {
    display: inline-block;
    vertical-align: middle;
  }
</style>

<docs>
  Component for svg icons

  ## Examples

  Eye icon in different sizes:

  ```jsx
  const eyeIcon = require('@/assets/icons/eye.svg').default;
  <div>
    <r-icon :glyph="eyeIcon" standalone size="small" />
    <r-icon :glyph="eyeIcon" size="medium" />
    <r-icon :glyph="eyeIcon" size="large" />
  </div>
  ```

  Standalone icon with aria label
  ```jsx
  const eyeIcon = require('@/assets/icons/eye.svg').default;
  <r-icon :glyph="eyeIcon" size="medium" standalone aria-label="eye icon" />
  ```

  Decorative icon without aria label
  ```jsx
  const eyeIcon = require('@/assets/icons/eye.svg').default;
  <r-icon :glyph="eyeIcon" size="medium" />
  ```
</docs>
