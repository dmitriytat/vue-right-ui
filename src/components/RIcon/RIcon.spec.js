import { shallowMount } from '@vue/test-utils';
import { createRenderer } from 'vue-server-renderer';

import RIcon from './RIcon.vue';

describe('RIcon', () => {
  const renderer = createRenderer();

  it('should render icon', () => {
    const glyph = {
      id: 'eye-usage',
      url: 'img/icons.svg#eye',
      viewBox: '0 0 18 18',
    };

    const wrapper = shallowMount(RIcon, {
      propsData: { glyph },
    });

    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err);

      expect(str).toMatchSnapshot();
    });
  });

  it('should render standalone icon with aria-label', () => {
    const glyph = {
      id: 'eye-usage',
      url: 'img/icons.svg#eye',
      viewBox: '0 0 18 18',
    };

    const wrapper = shallowMount(RIcon, {
      propsData: {
        glyph,
        standalone: true,
        ariaLabel: 'standalone icon',
      },
    });

    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err);

      expect(str).toMatchSnapshot();
    });
  });
});
