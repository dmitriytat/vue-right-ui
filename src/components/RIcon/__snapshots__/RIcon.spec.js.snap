// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RIcon should render icon 1`] = `
<svg width="12" height="12" viewBox="0 0 18 18" aria-hidden="true" class="r-icon">
  <use xlink:href="img/icons.svg#eye"></use>
</svg>
`;

exports[`RIcon should render standalone icon with aria-label 1`] = `
<svg width="12" height="12" viewBox="0 0 18 18" role="img" ariaLabel="standalone icon" class="r-icon">
  <use xlink:href="img/icons.svg#eye"></use>
</svg>
`;
